src/python/pants/backend/python/util_rules/lockfile_test.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from textwrap import dedent
+from typing import Iterable, Type
+
+from pants.backend.python.dependency_inference.rules import import_rules
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.backend.python.target_types import ConsoleScript
+from pants.backend.python.util_rules.lockfile import LockfileRules
+from pants.core.goals import generate_lockfiles
+from pants.core.goals.generate_lockfiles import GenerateLockfilesGoal, GenerateToolLockfileSentinel
+from pants.engine.target import Dependencies, SingleSourceField, Target
+from pants.engine.unions import UnionRule
+from pants.testutil.rule_runner import RuleRunner
+
+
+def _get_generated_lockfile_sentinel(
+    rules: Iterable, subsystem: Type[PythonToolBase]
+) -> Type[GenerateToolLockfileSentinel]:
+    """Fish the generated lockfile sentinel out of the pool of rules so it can be used in a
+    QueryRule."""
+    return next(
+        r
+        for r in rules
+        if isinstance(r, UnionRule)
+        and r.union_base == GenerateToolLockfileSentinel
+        and issubclass(r.union_member, GenerateToolLockfileSentinel)  # TypeGuard keeps mypy happy
+        and r.union_member.resolve_name == subsystem.options_scope
+    ).union_member
+
+
+class FakeTool(PythonToolBase):
```

Comment:
So the way we've been doing this now is to have the class' `.rules()` do the yielding, so it's a one-stop-shop.

So the design should be we override `.rules()` in `PythonToolBase`. We replace the `register_lockfile` classvar with `lockfile_rules_type` whose value is an enum, and the default is `CUSTOM` (e.g. no rules registered). Then in the overriden `rules` we look at the classvar and register accordingly.

It then becomes one less thing for the user to remember to do, and much easier to "get right".

LMK if you need some help on that design.

Reply:
I started with something like that, but I needed to shift away from that because some of the rules need references to other classes (like `from_tool_with_constraints` needing the `FieldSet`. Instead of an enum (value-based sum type) we could do a "sealed" class (class-based sum type), something like:
```python
class LockfileRulesType(ABC):
   @abstractmethod
    def rules(self, cls):
        ...

@dataclass
class LockfileFromToolWithConstraints(LockfileRulesType):
    field_set_type: Type[FieldSet]

    def rules(self, cls):
        ...  # implementation here
```
usage
```python
class PyTest(PythonToolBase):
    lockfile_rules_type = LockfileFromToolWithConstraints(PythonTestFieldSet)

    @memoized_classmethod
    def rules(cls: Any):
        parent_rules = super().rules(cls)
        lockfile_rules = cls.lockfile_rules_type.rules(cls)
```
thoughts?

Reply:
We haven't had to do anything too funny in the lint/fix/fmt core code so I think we can make it work here. I don't quite understand the stumbling block you ran into, can you help me see where this starts getting shaky and we can unblock it together?

Reply:
The gap we need to close is that some of these rules need to be aware of what FieldSet is "associated" with the tool. For a concrete example, `PyTest` is associated with  `PythonTestFieldSet`, and this allows us to pull `_find_all_unique_interpreter_constraints`.
Locating the associated fieldset is something that other rule generators need to do too. For example, for Partitions, `_single_partition_field_set_rules` needs to understand what fieldset it should be using. It does this by synthesising a rule that requests a `_PartitionFieldSetsRequestBase` which has a generic attr `field_sets: tuple[_FieldSetT, ...]` . Tools needing to use this partitioner need to subclass this. Usually they don't do that directly. `LintTargetsRequest` has an inner class 
```python
    @distinct_union_type_per_subclass(in_scope_types=[EnvironmentName])
    class PartitionRequest(_PartitionFieldSetsRequestBase[_FieldSetT]):
        pass
```
as well as the class attr `field_set_type: ClassVar[type[FieldSet]]`.
The `field_set_type` is finally set in the per-tool subclass of `LintTargetsRequest`:
```python
class HelmLintRequest(LintTargetsRequest):
    field_set_type = HelmLintFieldSet
    tool_subsystem = HelmSubsystem  # type: ignore[assignment]
```
Which is to say, the association between tool and FS is done manually in a class. 
---
If we wanted to do something similar, we could create a class `LockfileFromToolWithConstraints` analogous to `_PartitionFieldSetsRequestBase`, add that as a field in to GenerateToolLockfileSentinel (probably a more specific subclass), and have people create a tool-specific subclass of that.

I think the main difference between the approach of this MR and Partitions is that this MR closes the gap with an explicit function call; Partitions closes that gap through Goal-specific target requests. I think I did it this way so it would be just extracting existing code. I think the only _major_ UX change would be that we'd have people creating subclasses instead of invoking functions which generate those classes. Lemme run off an try that.

Reply:
Putting the rules generation on the goal-based subclass and having that class reference the associated tool is key to how Partition makes the rule generation and registration work with just the enum.
There are, at a minimum, 3 components necessary for to enable lockfiles (or similar) in the Pants engine: 
1. A subclass of the goal request class
2. a rule to go from that (and the tool class) to the general case
3. a UnionRule registering the request class
Having the rule generation happen on the goal-based subclass ensures that 1 exists and that we have a reference to it for 3. This does mean that we cannot autogenerate it...
Trying to implement the rule generation on PythonToolBase doesn't give us access to 1, so it needs a reference to it, which seems clumsy. It seems like we'd give it 2 class params, essentially which function to invoke and what arguments to pass to it. I think an explicit wiring for that would use less machinery and be overall clearer.
Moving it off of PythonToolBase means that we'd have to register the rules anyhow.

Reply:
I think one difference between fmt/fix/lint and export/lockfiles is that the first group are the capabilities of the plugin and the second group are life-support for the tool itself. It seems inevitable that you'd need to make `class AutoflakeRequest(FixTargetsRequest)`; how else does Pants know what autoflake does? and maybe the tool can do multiple things, so it makes sense to have separate `CheckRequest` and `FixRequest` and friends. But lockfiles are coupled to the tool and not the main goal of the plugin. (At least for the scope of this MR.)

Reply:
So, just looking at the code I don't see why the `PythonToolBase` (or whichever is the appropriate base class) can't export a classmethod `.rules()` which just `yield from LockfileRules.from_tool(cls)`?

Reply:
For the simple python case, sure. But for the other cases, they need to know which `FieldSet` type they need. For example, Bandit needs to get the interpreter constraints. To do that, we synthesis a rule which invokes `_find_all_unique_interpreter_constraints` which needs the FS type. I don't think it's possible to avoid this by using a more general FS type (one that has the `interpreter_constraints` field only, for example) since `_find_all_unique_interpreter_constraints` needs to invoke the FS's `opt_out` method which references the `SkipBanditField`, which is tool-specific.

This is why I think we need to find some way to close the gap of MyTool -> MyToolFieldSet. I don't know that a mechanism exists to do that just from the class MyTool, which is all we have from `PythonToolBase.rules()`. 

I've pushed some code which starts this work. Specifically where I'm stuck is in [here](https://github.com/pantsbuild/pants/pull/17480/files#diff-5b63f7e94d8ba371a5e5861b67abadd09ad2dcddea7c85ca2cfc2d412b651cf0R54), and how to get the  remaining arguments for `_pex_constraints_lockfile_rules`.

Reply:
Should PythonTools optionally declare a FieldSet classvar then if the synthesized rules needs it? That's kinda what we do for `lint` and friends (minus it being optional). The linked code would use `cls.field_set_type` and we'd assert (or a more friendly error) that its not `None`.

Reply:
I should add, there's probably good reasons to have the class declare what field set it will be using. Such as enforcing that it DOES contain an opt-out mechanism.

Reply:
(Marking as resolved)